src/arch/util/machine-smp.C

```diff
@@ -514,48 +512,117 @@ static void CmiDestroyLocks(void)
 
 #if !CMK_SHARED_VARS_UNAVAILABLE
 
-#if __cpp_lib_barrier
-
-using Barrier = std::barrier;
-
-#else
-
-class Barrier {
+class Barrier
+{
 public:
   Barrier(const Barrier&) = delete;
   Barrier& operator=(const Barrier&) = delete;
 
-  explicit Barrier(unsigned int count) :
-    curCount(count), barrierCount(count), curSense(true) {
+  explicit Barrier(unsigned int count)
+      : curCount(count), barrierCount(count), curSense(true)
+  {
   }
 
-  void arrive_and_wait() {
-    std::unique_lock<std::mutex> lock(barrierMutex);
-    const bool sense = curSense;
+  void arrive_and_wait()
+  {
+    const bool sense = curSense.load(std::memory_order_relaxed);
+
+    // If we're last, reset the count and flip the sense.
+    // Using acq_rel here effectively makes this read-modify-write a fence.
+    // Note: we compare against 1 because fetch_sub returns the previous value.
+    if (curCount.fetch_sub(1, std::memory_order_acq_rel) == 1)
+    {
+      curCount.store(barrierCount, std::memory_order_relaxed);
+      curSense.store(!sense, std::memory_order_release);
+#  if __cpp_lib_atomic_wait
+      curSense.notify_all();
```

Comment:
Doesn't that need `<condition_variable>` ?

Reply:
Nope, just `<atomic>` (https://en.cppreference.com/w/cpp/atomic/atomic/wait and https://en.cppreference.com/w/cpp/atomic/atomic/notify_all). It's a C++20 addition.